gunpowder/jax/nodes/train.py

```diff
@@ -0,0 +1,288 @@
+                                    self.model_params, inputs, False)
+
+        logger.debug(
+            "model outputs: %s",
+            {k: v.shape for k, v in outputs.items()})
+
+        # add requested model outputs to batch
+        for array_key, array_name in requested_outputs.items():
+            spec = self.spec[array_key].copy()
+            spec.roi = request[array_key].roi
+            batch.arrays[array_key] = Array(
+                outputs[array_name], spec
+            )
+
+        batch.loss = loss
+        self.iteration += 1
+        batch.iteration = self.iteration
+
+        if batch.iteration % self.save_every == 0:
+
+            checkpoint_name = self._checkpoint_name(
+                self.checkpoint_basename, batch.iteration)
+
+            logger.info("Creating checkpoint %s", checkpoint_name)
+
+            model_state = self.model_params
+            if self.n_devices > 1:
+                # get only a single copy of param for saving
+                model_state = jax.tree_map(lambda x: x[0], model_state)
+            with open(checkpoint_name, 'wb') as f:
+                pickle.dump(model_state, f)
+
+            if self.keep_n_checkpoints:
+                checkpoint_name = self._checkpoint_name(
+                    self.checkpoint_basename,
+                    batch.iteration - self.keep_n_checkpoints*self.save_every)
+                try:
+                    os.remove(checkpoint_name)
+                    logger.info("Removed checkpoint %s", checkpoint_name)
+                except FileNotFoundError:
+                    pass
+
+        if self.summary_writer and batch.iteration % self.log_every == 0:
+            self.summary_writer.add_scalar("loss", batch.loss, batch.iteration)
+
+    def __collect_requested_outputs(self, request):
+
+        array_outputs = {}
+
+        for output_name, array_key in self.outputs.items():
+            if array_key in request:
+                array_outputs[array_key] = output_name
+
+        return array_outputs
+
+    def __collect_provided_inputs(self, batch):
+
+        return self.__collect_provided_arrays(
+            {k: v for k, v in self.inputs.items()}, batch
+        )
+
+    def __collect_provided_arrays(self, reference, batch, expect_missing_arrays=False):
+
+        arrays = {}
+
+        for array_name, array_key in reference.items():
+            if isinstance(array_key, ArrayKey):
+                msg = f"batch does not contain {array_key}, array {array_name} will not be set"
+                if array_key in batch.arrays:
+                    arrays[array_name] = batch.arrays[array_key].data
+                elif not expect_missing_arrays:
+                    logger.warn(msg)
+                else:
+                    logger.debug(msg)
+            elif isinstance(array_key, np.ndarray):
+                arrays[array_name] = array_key
+            elif isinstance(array_key, str):
+                arrays[array_name] = getattr(batch, array_key)
```

Comment:
Should avoid setting string attributes on batches. We've used "loss" and "iteration" like this but that should change and we should avoid allowing any more support for this.
Mainly due to the fact that the gunpowder pipeline often needs to merge 2 batches or create a new batch with a subset of the data in batch. Both of these operations are more annoying with arbitrary string attributes.

Reply:
Yea I've never used it either; I can remove line 280-281